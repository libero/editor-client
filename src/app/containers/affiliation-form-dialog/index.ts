export * from './affiliation-form-dialog';
export * from './connected-affiliation-form-dialog';
