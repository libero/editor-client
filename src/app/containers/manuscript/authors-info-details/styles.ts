import { makeStyles } from '@material-ui/core/styles';

export const useAuthorsDetailsListStyles = makeStyles((theme) => ({
  addAuthorButton: {
    marginLeft: theme.spacing(1)
  }
}));

export const useAuthorDetailStyles = makeStyles((theme) => ({
  root: {
    display: 'flex',
    alignItems: 'flex-start',
    marginTop: theme.spacing(3),
    '&:first-of-type': {
      marginTop: theme.spacing(0)
    }
  },
  editButton: {
    flexGrow: 0,
    padding: 6
  },
  authorInfoContainer: {
    flex: 1
  },
  authorInfoLine: {
    lineHeight: '24px'
  },
  authorBio: {
    lineHeight: '24px',
    marginTop: theme.spacing(1.5)
  },
  orcidIcon: {
    width: 16,
    height: 16,
    marginTop: 4,
    display: 'inline',
    verticalAlign: 'top'
  }
}));
