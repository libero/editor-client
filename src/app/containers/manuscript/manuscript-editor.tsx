import React, { useCallback } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { EditorState, Transaction } from 'prosemirror-state';

import { getAbstract, getKeywordGroups, getTitle } from 'app/selectors/manuscript.selectors';
import * as manuscriptActions from 'app/actions/manuscript.actions';
import { RichTextEditor } from 'app/components/rich-text-editor';
import * as manuscriptEditorActions from 'app/actions/manuscript-editor.actions';
import { KeywordsEditor } from 'app/components/keywords';
import { KeywordGroups } from 'app/models/manuscript';
import { useManuscriptStyles } from './styles';
import { SortableAuthorsList } from './sortable-authors-list';
import { AuthorsInfoDetails } from 'app/containers/manuscript/authors-info-details';
import { AffiliationsList } from 'app/containers/manuscript/affiliations-list';
import { ReferenceList } from 'app/containers/manuscript/references-list';
import { ArticleInformation } from 'app/containers/manuscript/article-information';
import { getFocusedEditorStatePath } from 'app/selectors/manuscript-editor.selectors';

const isInputFocused = (inputName: string, focusedPath?: string) => {
  return Boolean(focusedPath) && focusedPath.startsWith(inputName);
};

export const ManuscriptEditor: React.FC = () => {
  const classes = useManuscriptStyles();
  const dispatch = useDispatch();

  const title: EditorState = useSelector(getTitle);
  const abstract: EditorState = useSelector(getAbstract);
  const focusedPath = useSelector(getFocusedEditorStatePath);
  const allKeywords: KeywordGroups = useSelector(getKeywordGroups);

  const handleTitleChange = useCallback(
    (diff: Transaction) => {
      dispatch(manuscriptActions.updateTitleAction(diff));
    },
    [dispatch]
  );

  const handleKeywordsChange = (keywordGroup: string, index: number, diff: Transaction): void => {
    dispatch(manuscriptActions.updateKeywordAction({ keywordGroup, index, change: diff }));
  };

  const handleNewKeywordChange = (keywordGroup: string, diff: Transaction): void => {
    dispatch(manuscriptActions.updateNewKeywordAction({ keywordGroup, change: diff }));
  };

  const handleAbstractChange = useCallback(
    (diff: Transaction): void => {
      dispatch(manuscriptActions.updateAbstractAction(diff));
    },
    [dispatch]
  );

  const handleKeywordDelete = (keywordGroup: string, index: number): void => {
    dispatch(manuscriptActions.deleteKeywordAction({ keywordGroup, index }));
  };

  const handleKeywordAdd = (keywordGroup: string, keyword: EditorState): void => {
    dispatch(manuscriptActions.addNewKeywordAction({ keywordGroup, keyword }));
  };

  const handleFocusSwitch = useCallback(
    (state: EditorState, path: string) => {
      dispatch(manuscriptEditorActions.setFocusAction(path));
    },
    [dispatch]
  );

  const handleBlur = useCallback(() => {
    dispatch(manuscriptEditorActions.removeFocusAction());
  }, [dispatch]);

  const handleKeywordFocus = (group: string, index: number | undefined, isNewKeywordFocused: boolean): void => {
    const kwdIndexPath = isNewKeywordFocused ? 'newKeyword' : `keywords.${index}`;
    handleFocusSwitch(null, ['keywordGroups', group, kwdIndexPath].join('.'));
  };

  const renderKeywords = (keywordGroups: KeywordGroups): JSX.Element[] => {
    return Object.entries(keywordGroups).map(([groupType, group]) => {
      return (
        <KeywordsEditor
          key={groupType}
          keywords={group.keywords}
          newKeyword={group.newKeyword}
          activeKeywordPath={focusedPath}
          name={groupType}
          label={group.title || 'Keywords'}
          onNewKeywordChange={handleNewKeywordChange}
          onAdd={handleKeywordAdd}
          onChange={handleKeywordsChange}
          onDelete={handleKeywordDelete}
          onFocusSwitch={handleKeywordFocus}
          onBlur={handleBlur}
        />
      );
    });
  };

  return (
    <div className={classes.content}>
      <div aria-hidden="true" className={classes.toolbarPlaceholder} />
      <RichTextEditor
        editorState={title}
        label="Title"
        isActive={isInputFocused('title', focusedPath)}
        name="title"
        onChange={handleTitleChange}
        onFocusSwitch={handleFocusSwitch}
      />
      <SortableAuthorsList />
      <AffiliationsList />
      <RichTextEditor
        editorState={abstract}
        label="Abstract"
        name="abstract"
        isActive={isInputFocused('abstract', focusedPath)}
        onChange={handleAbstractChange}
        onFocusSwitch={handleFocusSwitch}
      />
      {renderKeywords(allKeywords)}
      <ArticleInformation />
      <AuthorsInfoDetails />
      <ReferenceList />
    </div>
  );
};
