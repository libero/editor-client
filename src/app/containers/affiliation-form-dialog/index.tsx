import React, { SyntheticEvent, useCallback, useState } from 'react';
import { TextField } from '@material-ui/core';
import { useDispatch, useSelector } from 'react-redux';
import { set, cloneDeep } from 'lodash';

import { useAffiliationFormStyles } from './styles';
import * as manuscriptEditorActions from 'app/actions/manuscript-editor.actions';
import * as manuscriptActions from 'app/actions/manuscript.actions';
import { PromptDialog } from 'app/components/prompt-dialog';
import { ActionButton } from 'app/components/action-button';
import { Affiliation, createAffiliation } from 'app/models/affiliation';
import { getAffiliatedAuthors, getAuthors } from 'app/selectors/manuscript.selectors';
import { LinkedAuthorsList } from 'app/containers/affiliation-form-dialog/linked-authors-list';
import { Person } from 'app/models/person';
import { linkAffiliationsAction } from 'app/actions/manuscript.actions';

interface AffiliationFormDialogProps {
  affiliation?: Affiliation;
}

const renderConfirmDialog = (title: string, msg: string, onAccept: () => void, onReject: () => void) => {
  return (
    <PromptDialog
      title={title}
      message={msg}
      isOpen={true}
      onAccept={onAccept}
      onReject={onReject}
      acceptLabel="Delete"
      rejectLabel="Cancel"
      acceptVariant="containedWarning"
      rejectVariant="secondaryOutlined"
    />
  );
};

const labelProps = { shrink: true };

export const AffiliationFormDialog: React.FC<AffiliationFormDialogProps> = (props) => {
  const dispatch = useDispatch();
  const isNewAffiliation = !props.affiliation;

  const affiliatedAuthors = useSelector(getAffiliatedAuthors)(props.affiliation?.id);
  const [userLinkedAuthors, setUserLinkedAuthors] = useState<Person[]>(affiliatedAuthors);
  const allAuthors = useSelector(getAuthors);

  const [affiliation, setAffiliation] = useState<Affiliation>(
    props.affiliation ||
      createAffiliation(undefined, {
        label: '',
        institution: {
          department: '',
          name: ''
        },
        address: {
          city: ''
        },
        country: ''
      })
  );
  const [isConfirmShown, setConfirmShow] = useState<boolean>(false);

  const classes = useAffiliationFormStyles();

  const closeDialog = useCallback(() => {
    dispatch(manuscriptEditorActions.hideModalDialog());
  }, [dispatch]);

  const handleDelete = useCallback(() => {
    setConfirmShow(true);
  }, [setConfirmShow]);

  const handleAccept = useCallback(() => {
    setConfirmShow(false);
    dispatch(manuscriptActions.deleteAffiliationAction(affiliation));
    closeDialog();
  }, [setConfirmShow, affiliation, closeDialog, dispatch]);

  const handleReject = useCallback(() => {
    setConfirmShow(false);
  }, [setConfirmShow]);

  const handleLinkedAuthorsChange = useCallback((authors: Person[]) => {
    setUserLinkedAuthors(authors);
  }, []);

  const handleFormChange = useCallback(
    (event: SyntheticEvent) => {
      const fieldName = event.target['name'];
      const newValue = event.target['value'];
      const newAffiliation = set(cloneDeep(affiliation), fieldName, newValue);
      setAffiliation(newAffiliation);
    },
    [affiliation, setAffiliation]
  );

  const handleDone = useCallback(() => {
    if (isNewAffiliation) {
      dispatch(manuscriptActions.addAffiliationAction(affiliation));
    } else {
      dispatch(manuscriptActions.updateAffiliationAction(affiliation));
    }
    dispatch(linkAffiliationsAction({ affiliation, authors: userLinkedAuthors }));
    closeDialog();
  }, [affiliation, closeDialog, dispatch, isNewAffiliation, userLinkedAuthors]);

  return (
    <section className={classes.root}>
      <TextField
        fullWidth
        name="institution.department"
        label="Department"
        classes={{ root: classes.inputField }}
        InputLabelProps={labelProps}
        variant="outlined"
        value={affiliation.institution.department}
        onChange={handleFormChange}
      />
      <TextField
        fullWidth
        name="institution.name"
        label="Institute"
        classes={{ root: classes.inputField }}
        InputLabelProps={labelProps}
        variant="outlined"
        value={affiliation.institution.name}
        onChange={handleFormChange}
      />
      <TextField
        name="address.city"
        label="City"
        classes={{ root: classes.inputField }}
        InputLabelProps={labelProps}
        variant="outlined"
        value={affiliation.address.city}
        onChange={handleFormChange}
      />
      <TextField
        name="country"
        fullWidth
        label="Country"
        classes={{ root: classes.inputField }}
        InputLabelProps={labelProps}
        variant="outlined"
        value={affiliation.country}
        onChange={handleFormChange}
      />
      <LinkedAuthorsList
        linkedAuthors={userLinkedAuthors}
        allAuthors={allAuthors}
        onChange={handleLinkedAuthorsChange}
      />
      <div className={classes.buttonPanel}>
        {!isNewAffiliation ? (
          <ActionButton variant="outlinedWarning" onClick={handleDelete} title="Delete" />
        ) : undefined}
        <div aria-hidden={true} className={classes.spacer}></div>
        <ActionButton variant="secondaryOutlined" onClick={closeDialog} title="Cancel" />
        <ActionButton variant="primaryContained" onClick={handleDone} title="Done" />
      </div>
      {isConfirmShown
        ? renderConfirmDialog(
            'You are deleting an affiliation',
            'Deleting an affiliation will remove all associated links with authors. Are you sure want to proceed?',
            handleAccept,
            handleReject
          )
        : undefined}
    </section>
  );
};
