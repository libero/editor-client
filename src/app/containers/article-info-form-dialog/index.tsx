import React, { useCallback, SyntheticEvent, useState, ChangeEvent } from 'react';
import { TextField } from '@material-ui/core';
import { useDispatch, useSelector } from 'react-redux';
import { isEqual, set } from 'lodash';
import Interweave from 'interweave';

import * as manuscriptEditorActions from 'app/actions/manuscript-editor.actions';
import * as manuscriptActions from 'app/actions/manuscript.actions';
import { ActionButton } from 'app/components/action-button';
import { useArticleInfoFormStyles } from 'app/containers/article-info-form-dialog/styles';
import { getArticleInformation, getAuthors } from 'app/selectors/manuscript.selectors';
import formGrid from 'app/styles/form-grid.module.scss';
import { Select } from 'app/components/select';
import {
  ArticleInformation,
  getCopyrightStatement,
  getLicenseTextEditorState,
  LICENSE_CC0,
  LICENSE_CC_BY_4
} from 'app/models/article-information';
import { SectionContainer } from 'app/components/section-container';
import { stringifyEditorState } from 'app/utils/view.utils';

const labelProps = { shrink: true };

const SubjectOptions = [
  'Biochemistry and Chemical Biology',
  'Cancer Biology',
  'Cell Biology',
  'Chromosomes and Gene Expression',
  'Computational and Systems Biology',
  'Developmental Biology',
  'Ecology',
  'Epidemiology and Global Health',
  'Evolutionary Biology',
  'Genetics and Genomics',
  'Medicine',
  'Immunology and Inflammation',
  'Microbiology and Infectious Disease',
  'Neuroscience',
  'Physics of Living Systems',
  'Plant Biology',
  'Stem Cells and Regenerative Medicine',
  'Structural Biology and Molecular Biophysics'
].map((title) => ({ label: title, value: title }));

export const ArticleInfoFormDialog: React.FC<{}> = () => {
  const classes = useArticleInfoFormStyles();
  const dispatch = useDispatch();
  const articleInfo = useSelector(getArticleInformation);

  const [userArticleInfo, setArticleInfo] = useState<ArticleInformation>(articleInfo);
  const authors = useSelector(getAuthors);
  const closeDialog = useCallback(() => {
    dispatch(manuscriptEditorActions.hideModalDialog());
  }, [dispatch]);

  const handleFormChange = useCallback(
    (event: SyntheticEvent) => {
      const fieldName = event.target['name'];
      const newValue = event.target['value'];
      setArticleInfo({
        ...userArticleInfo,
        [fieldName]: newValue
      });
    },
    [userArticleInfo, setArticleInfo]
  );

  const handleSubjectChange = useCallback(
    (event: ChangeEvent<{ name: string; value: string }>) => {
      const fieldName = event.target['name'];
      const newValue = event.target['value'];
      const updatedArticleInfo = {
        ...userArticleInfo,
        subjects: [...userArticleInfo.subjects]
      };

      set(updatedArticleInfo, fieldName, newValue);
      updatedArticleInfo.subjects = updatedArticleInfo.subjects.filter(Boolean);
      setArticleInfo(updatedArticleInfo);
    },
    [userArticleInfo, setArticleInfo]
  );

  const handleLicenseTypeChange = useCallback(
    (event: ChangeEvent<{ name: string; value: string }>) => {
      const newLicenseType = event.target['value'];
      const updatedArticleInfo = {
        ...userArticleInfo,
        licenseType: newLicenseType,
        licenseText: getLicenseTextEditorState(newLicenseType),
        copyrightStatement:
          newLicenseType === LICENSE_CC_BY_4 ? getCopyrightStatement(authors, userArticleInfo.publicationDate) : ''
      };
      setArticleInfo(updatedArticleInfo);
    },
    [userArticleInfo, authors]
  );

  const handleAccept = useCallback(() => {
    if (!isEqual(userArticleInfo, articleInfo)) {
      dispatch(manuscriptActions.updateArticleInformationAction(userArticleInfo));
    }
    closeDialog();
  }, [closeDialog, articleInfo, userArticleInfo, dispatch]);

  return (
    <section className={classes.root}>
      <div className={formGrid.container}>
        <Select
          className={formGrid.fullWidth}
          name="subjects.0"
          placeholder="Please select"
          fullWidth
          canUnselect={true}
          blankValue={undefined}
          label="Subject *"
          value={userArticleInfo.subjects[0]}
          onChange={handleSubjectChange}
          options={SubjectOptions}
        />
        <Select
          className={formGrid.fullWidth}
          name="subjects.1"
          placeholder="Please select"
          fullWidth
          canUnselect={true}
          blankValue={undefined}
          label="Subject"
          value={userArticleInfo.subjects[1]}
          onChange={handleSubjectChange}
          options={SubjectOptions}
        />
        <TextField
          fullWidth
          name="articleDOI"
          label="Article DOI"
          classes={{ root: formGrid.fullWidth }}
          InputLabelProps={labelProps}
          variant="outlined"
          value={userArticleInfo.articleDOI}
          onChange={handleFormChange}
        />
        <TextField
          fullWidth
          name="publisherId"
          label="Publisher ID"
          classes={{ root: formGrid.fullWidth }}
          InputLabelProps={labelProps}
          variant="outlined"
          value={userArticleInfo.publisherId}
          onChange={handleFormChange}
        />
        <TextField
          fullWidth
          name="elocationId"
          label="eLocation ID"
          classes={{ root: formGrid.fullWidth }}
          InputLabelProps={labelProps}
          variant="outlined"
          value={userArticleInfo.elocationId}
          onChange={handleFormChange}
        />
        <TextField
          fullWidth
          name="volume"
          label="Volume"
          classes={{ root: formGrid.firstCol }}
          InputLabelProps={labelProps}
          variant="outlined"
          value={userArticleInfo.volume}
          onChange={handleFormChange}
        />
        <TextField
          fullWidth
          name="publicationDate"
          label="Published"
          type="date"
          classes={{ root: formGrid.fullWidth }}
          InputLabelProps={labelProps}
          variant="outlined"
          value={userArticleInfo.publicationDate}
          onChange={handleFormChange}
        />
        <Select
          className={formGrid.fullWidth}
          placeholder="Please select"
          fullWidth
          blankValue={undefined}
          label="License type"
          value={userArticleInfo.licenseType}
          onChange={handleLicenseTypeChange}
          options={[
            { label: LICENSE_CC_BY_4, value: LICENSE_CC_BY_4 },
            { label: LICENSE_CC0, value: LICENSE_CC0 }
          ]}
        />
        {userArticleInfo.licenseType === LICENSE_CC_BY_4 ? (
          <SectionContainer label="Copyright statement" variant="outlined" className={formGrid.fullWidth}>
            {userArticleInfo.copyrightStatement}
          </SectionContainer>
        ) : undefined}
        <SectionContainer label="Permissions" variant="outlined" className={formGrid.fullWidth}>
          <Interweave content={stringifyEditorState(userArticleInfo.licenseText)} />
        </SectionContainer>
      </div>
      <div className={classes.buttonPanel}>
        <ActionButton variant="secondaryOutlined" onClick={closeDialog} title="Cancel" />
        <ActionButton variant="primaryContained" onClick={handleAccept} title="Done" />
      </div>
    </section>
  );
};
