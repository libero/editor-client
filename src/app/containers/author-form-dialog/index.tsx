import React, { SyntheticEvent, useCallback, useState } from 'react';
import { TextField } from '@material-ui/core';
import { useDispatch, useSelector } from 'react-redux';

import { useAuthorFormStyles } from './styles';
import { createAuthor, Person } from 'app/models/person';
import * as manuscriptEditorActions from 'app/actions/manuscript-editor.actions';
import * as manuscriptActions from 'app/actions/manuscript.actions';
import { PromptDialog } from 'app/components/prompt-dialog';
import { ActionButton } from 'app/components/action-button';
import { LinkedAffiliationsList } from './linked-affiliations-list';
import { getAffiliations, getAuthorAffiliations } from 'app/selectors/manuscript.selectors';
import { Affiliation } from 'app/models/affiliation';

interface AuthorFormDialogProps {
  author?: Person;
}

const renderConfirmDialog = (title: string, msg: string, onAccept: () => void, onReject: () => void) => {
  return (
    <PromptDialog
      title={title}
      message={msg}
      isOpen={true}
      onAccept={onAccept}
      onReject={onReject}
      acceptLabel="Delete"
      rejectLabel="Cancel"
      acceptVariant="containedWarning"
      rejectVariant="secondaryOutlined"
    />
  );
};

const labelProps = { shrink: true };

export const AuthorFormDialog: React.FC<AuthorFormDialogProps> = (props) => {
  const dispatch = useDispatch();
  const isNewAuthor = !props.author;

  const [author, setAuthor] = useState<Person>(
    props.author || createAuthor(undefined, { firstName: '', lastName: '' })
  );
  const [isConfirmShown, setConfirmSnow] = useState<boolean>(false);

  const allAffiliations = useSelector(getAffiliations);
  const linkedAffiliations = useSelector(getAuthorAffiliations)(author);
  const classes = useAuthorFormStyles();

  const closeDialog = useCallback(() => {
    dispatch(manuscriptEditorActions.hideModalDialog());
  }, [dispatch]);

  const handleDelete = useCallback(() => {
    setConfirmSnow(true);
  }, [setConfirmSnow]);

  const handleAccept = useCallback(() => {
    setConfirmSnow(false);
    dispatch(manuscriptActions.deleteAuthorAction(author));
    closeDialog();
  }, [setConfirmSnow, author, closeDialog, dispatch]);

  const handleAffiliationsUpdate = useCallback(
    (affiliations: Affiliation[]) => {
      author.affiliations = affiliations.map(({ id }) => id);
    },
    [author]
  );

  const handleReject = useCallback(() => {
    setConfirmSnow(false);
  }, [setConfirmSnow]);

  const handleFormChange = useCallback(
    (event: SyntheticEvent) => {
      const fieldName = event.target['name'];
      const newValue = event.target['value'];
      setAuthor({
        ...author,
        [fieldName]: newValue
      });
    },
    [author, setAuthor]
  );

  const handleDone = useCallback(() => {
    if (isNewAuthor) {
      dispatch(manuscriptActions.addAuthorAction(author));
    } else {
      dispatch(manuscriptActions.updateAuthorAction(author));
    }
    closeDialog();
  }, [author, closeDialog, dispatch, isNewAuthor]);

  return (
    <section className={classes.root}>
      <TextField
        fullWidth
        name="firstName"
        label="Given name(s)"
        classes={{ root: classes.inputField }}
        InputLabelProps={labelProps}
        variant="outlined"
        value={author.firstName}
        onChange={handleFormChange}
      />
      <TextField
        fullWidth
        name="lastName"
        label="Surname(s)"
        classes={{ root: classes.inputField }}
        InputLabelProps={labelProps}
        variant="outlined"
        value={author.lastName}
        onChange={handleFormChange}
      />
      <TextField
        name="suffix"
        label="Suffix"
        classes={{ root: classes.inputField }}
        InputLabelProps={labelProps}
        variant="outlined"
        value={author.suffix || ''}
        onChange={handleFormChange}
      />
      <TextField
        name="email"
        fullWidth
        label="Email"
        classes={{ root: classes.inputField }}
        InputLabelProps={labelProps}
        variant="outlined"
        value={author.email || ''}
        onChange={handleFormChange}
      />
      <LinkedAffiliationsList
        allAffiliations={allAffiliations}
        linkedAffiliations={linkedAffiliations}
        onChange={handleAffiliationsUpdate}
      />
      <div className={classes.buttonPanel}>
        {!isNewAuthor ? <ActionButton variant="outlinedWarning" onClick={handleDelete} title="Delete" /> : undefined}
        <div aria-hidden={true} className={classes.spacer}></div>
        <ActionButton variant="secondaryOutlined" onClick={closeDialog} title="Cancel" />
        <ActionButton variant="primaryContained" onClick={handleDone} title="Done" />
      </div>
      {isConfirmShown
        ? renderConfirmDialog(
            'You are deleting an author',
            'Deleting an author can leave unlinked affiliation and funding information. Are you sure you want to proceed?',
            handleAccept,
            handleReject
          )
        : undefined}
    </section>
  );
};
