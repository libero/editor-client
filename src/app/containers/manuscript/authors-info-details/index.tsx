import React, { useCallback } from 'react';
import { IconButton } from '@material-ui/core';
import { useSelector, useDispatch } from 'react-redux';
import EditIcon from '@material-ui/icons/Edit';
import { isBoolean, isEqual } from 'lodash';
import Interweave from 'interweave';

import { SectionContainer } from 'app/components/section-container';
import { getAuthorAffiliations, getAuthors } from 'app/selectors/manuscript.selectors';
import { Person } from 'app/models/person';
import {
  useAuthorDetailStyles,
  useAuthorsDetailsListStyles
} from 'app/containers/manuscript/authors-info-details/styles';
import { ActionButton } from 'app/components/action-button';
import * as manuscriptEditorActions from 'app/actions/manuscript-editor.actions';
import { AuthorFormDialog } from 'app/containers/author-form-dialog';
import { stringifyEditorState } from 'app/utils/view.utils';
import { OrcidIcon } from 'app/assets/icons';
import { ComponentWithId } from 'app/types/utility.types';

export const AuthorsInfoDetails: React.FC<ComponentWithId> = ({ id }) => {
  const authors = useSelector(getAuthors);
  const classes = useAuthorsDetailsListStyles();
  const dispatch = useDispatch();

  const addAuthor = useCallback(() => {
    dispatch(
      manuscriptEditorActions.showModalDialog({
        component: AuthorFormDialog,
        title: 'Add Author'
      })
    );
  }, [dispatch]);

  return (
    <section>
      <SectionContainer label="Author Information" id={id}>
        {authors.map((author) => (
          <AuthorInfo key={author.id} author={author} />
        ))}
      </SectionContainer>
      <ActionButton variant="addEntity" title="Add Author" onClick={addAuthor} className={classes.addAuthorButton} />
    </section>
  );
};

interface AuthorInfoProps {
  author: Person;
}

const AuthorInfo: React.FC<AuthorInfoProps> = React.memo(({ author }) => {
  const classes = useAuthorDetailStyles();
  const dispatch = useDispatch();
  const affiliations = useSelector(getAuthorAffiliations)(author);

  const editAuthor = useCallback(() => {
    dispatch(
      manuscriptEditorActions.showModalDialog({
        component: AuthorFormDialog,
        props: { author },
        title: 'Edit Author'
      })
    );
  }, [author, dispatch]);

  return (
    <section className={classes.root}>
      <div className={classes.authorInfoContainer}>
        <div className={classes.authorInfoLine}>
          <strong> {author.getDisplayName()} </strong>
        </div>
        <div className={classes.authorInfoLine}>
          {author.isCorrespondingAuthor ? 'Corresponding Author: ' : undefined}
          {author.email}
        </div>
        <div className={classes.authorInfoLine}>
          {affiliations.map((aff) => (
            <div key={aff.id}>{aff.getDisplayName()}</div>
          ))}
        </div>
        <div className={classes.authorInfoLine}>
          {author.isAuthenticated ? <OrcidIcon className={classes.orcidIcon} /> : undefined}{' '}
          <a href={`https://orcid.org/${author.orcid}`} target="_blank" rel="noopener noreferrer">
            {author.orcid}
          </a>
        </div>
        {isBoolean(author.hasCompetingInterest) ? (
          <div className={classes.authorInfoLine}>
            {author.hasCompetingInterest
              ? `Competing interest: ${author.competingInterestStatement}`
              : 'No Competing interest declared'}
          </div>
        ) : undefined}
        {author.bio ? (
          <div className={classes.authorBio}>
            <Interweave content={stringifyEditorState(author.bio)} />
          </div>
        ) : undefined}
      </div>
      <IconButton classes={{ root: classes.editButton }} onClick={editAuthor}>
        <EditIcon fontSize="small" />
      </IconButton>
    </section>
  );
}, isEqual);
