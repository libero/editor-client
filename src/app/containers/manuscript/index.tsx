import React from 'react';
import { useSelector } from 'react-redux';
import { Backdrop, CircularProgress } from '@material-ui/core';

import { isManuscriptLoaded } from '../../selectors/manuscript.selectors';
import './styles.scss';
import { ManuscriptToolbar } from './manuscript-toolbar';
import { ManuscriptEditor } from './manuscript-editor';

const renderBackdrop = (): JSX.Element => (
  <Backdrop open={true}>
    <CircularProgress color="inherit" />
  </Backdrop>
);

export const ManuscriptContainer: React.FC = () => {
  const isLoaded = useSelector(isManuscriptLoaded);
  const renderContent = (): JSX.Element => (
    <React.Fragment>
      <ManuscriptToolbar />
      <ManuscriptEditor />
    </React.Fragment>
  );

  return isLoaded ? renderContent() : renderBackdrop();
};
