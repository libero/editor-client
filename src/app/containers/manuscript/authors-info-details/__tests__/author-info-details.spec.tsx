import React from 'react';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';

import { AuthorsInfoDetails } from 'app/containers/manuscript/authors-info-details/index';
import { getInitialHistory, getLoadableStateSuccess } from 'app/utils/state.utils';
import { EditorState } from 'prosemirror-state';
import { create } from 'react-test-renderer';
import { mount } from 'enzyme';
import * as manuscriptEditorActions from 'app/actions/manuscript-editor.actions';
import { AuthorFormDialog } from 'app/containers/author-form-dialog';
import { IconButton } from '@material-ui/core';
import { ActionButton } from 'app/components/action-button';

jest.mock('@material-ui/core', () => ({
  Button: ({ label }) => <div data-cmp="Button">{label}</div>,
  IconButton: () => <div data-cmp="IconButton"></div>
}));

jest.mock('app/components/rich-text-input', () => ({
  RichTextInput: () => <div data-cmp="rich-text-input"></div>
}));

const AUTHORS = [
  {
    id: '4d53e405-5225-4858-a87a-aec902ae50b6',
    firstName: 'Fred',
    lastName: 'Atherden'
  },
  {
    id: 'c3b008e6-4ae9-4ef9-b7cb-854749a1e897',
    firstName: 'Jeanine',
    lastName: 'Smith'
  }
];

describe('Authors info details', () => {
  const mockStore = configureMockStore([]);

  it('renders authors list', () => {
    const mockState = getInitialHistory({
      title: new EditorState(),
      abstract: new EditorState(),
      affiliations: [],
      keywordGroups: {},
      authors: AUTHORS,
      references: []
    });
    const store = mockStore({
      manuscript: getLoadableStateSuccess(mockState)
    });
    const wrapper = create(
      <Provider store={store}>
        <AuthorsInfoDetails />
      </Provider>
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('dispatches edit action', () => {
    const mockState = getInitialHistory({
      title: new EditorState(),
      affiliations: [],
      abstract: new EditorState(),
      keywordGroups: {},
      authors: AUTHORS,
      references: []
    });

    const store = mockStore({
      manuscript: getLoadableStateSuccess(mockState)
    });

    jest.spyOn(store, 'dispatch');

    const wrapper = mount(
      <Provider store={store}>
        <AuthorsInfoDetails />
      </Provider>
    );

    wrapper.find(IconButton).at(0).prop('onClick')(null);

    expect(store.dispatch).toBeCalledWith(
      manuscriptEditorActions.showModalDialog({
        component: AuthorFormDialog,
        props: { author: AUTHORS[0] },
        title: 'Edit Author'
      })
    );
  });

  it('dispatches add action', () => {
    const mockState = getInitialHistory({
      title: new EditorState(),
      abstract: new EditorState(),
      affiliations: [],
      keywordGroups: {},
      authors: [],
      references: []
    });

    const store = mockStore({
      manuscript: getLoadableStateSuccess(mockState)
    });

    jest.spyOn(store, 'dispatch');

    const wrapper = mount(
      <Provider store={store}>
        <AuthorsInfoDetails />
      </Provider>
    );

    wrapper.find(ActionButton).prop('onClick')();

    expect(store.dispatch).toBeCalledWith(
      manuscriptEditorActions.showModalDialog({
        component: AuthorFormDialog,
        title: 'Add Author'
      })
    );
  });
});
