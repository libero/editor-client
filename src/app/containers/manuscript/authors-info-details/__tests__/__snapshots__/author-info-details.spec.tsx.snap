// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Authors info details renders authors list 1`] = `
<section>
  <div
    className="makeStyles-root-2 makeStyles-plainVariant-4"
  >
    <label
      className="makeStyles-label-5"
    >
      Author Information
    </label>
    <section
      className="makeStyles-root-7"
    >
      <div>
        <div
          className="makeStyles-authorInfoLine-9"
        >
          <strong>
             
            Fred Atherden
             
          </strong>
        </div>
        <div
          className="makeStyles-authorInfoLine-9"
        >
          Corresponding Author: 
          test@example.com
        </div>
        <div
          className="makeStyles-authorInfoLine-9"
        />
        <div
          className="makeStyles-authorInfoLine-9"
        >
           
          <a
            href="https://orcid.org/undefined"
            rel="noopener noreferrer"
            target="_blank"
          />
        </div>
        <div
          className="makeStyles-authorInfoLine-9"
        >
          Competing interest: Works at eLife
        </div>
      </div>
      <div
        data-cmp="IconButton"
      />
    </section>
    <section
      className="makeStyles-root-7"
    >
      <div>
        <div
          className="makeStyles-authorInfoLine-9"
        >
          <strong>
             
            Jeanine Smith
             
          </strong>
        </div>
        <div
          className="makeStyles-authorInfoLine-9"
        />
        <div
          className="makeStyles-authorInfoLine-9"
        />
        <div
          className="makeStyles-authorInfoLine-9"
        >
           
          <a
            href="https://orcid.org/undefined"
            rel="noopener noreferrer"
            target="_blank"
          />
        </div>
        <div
          className="makeStyles-authorInfoLine-9"
        >
          No Competing interest declared
        </div>
      </div>
      <div
        data-cmp="IconButton"
      />
    </section>
  </div>
  <div
    data-cmp="Button"
  />
</section>
`;
